### PR TITLE
[clang] Diagnose dangling references for parenthesized aggregate initialization.

### LINKED ISSUE #101957: Detect dangling references for C++20's Parenthesized Aggregate Initialization
```cpp
struct MyThing {
    const std::string& s;
};

MyThing makeThing(std::string s) {
    return MyThing(s);
}
```
https://godbolt.org/z/6335YcKMM

This also doesn't work for `gsl::Pointer` types like `string_view` instead of `const std::string&`. https://godbolt.org/z/8bcaxTqjY

This is different from https://github.com/llvm/llvm-project/issues/58985. This still only needs stmt-local analysis and has probably a simple fix.

### PR DESCRIPTION
**Pull request by hokein**
Unlike brace initialization,  the parenthesized aggregate initialization in C++20 does not extend the lifetime of a temporary object bound to a reference in an aggreate. This can lead to dangling references:

```
struct A { const int& r; };
 A a1(1); // well-formed, but results in a dangling reference.
 ``` 

With this patch, clang will diagnose this common dangling issues.

Fixes #101957

### CODE REVIEW

**Review by Xazax-hun — APPROVED** [review]
LGTM, could you add a summary of the change to the commit message? I know the context is available in the linked issue, but I'd like the gist of it to be preserved in the commit history directly.

File: clang/test/SemaCXX/paren-list-agg-init.cpp [hunk 1]
```diff
@@ -116,6 +116,7 @@ void foo(int n) { // expected-note {{declared here}}
   B b2(A(1), {}, 1);
   // beforecxx20-warning@-1 {{aggregate initialization of type 'A' from a parenthesized list of values is a C++20 extension}}
   // beforecxx20-warning@-2 {{aggregate initialization of type 'B' from a parenthesized list of values is a C++20 extension}}
+  // expected-warning@-3 {{temporary whose address is used as value of local variable 'b2' will be destroyed at the end of the full-expression}}
```
Comment (usx95) [inline clang/test/SemaCXX/paren-list-agg-init.cpp:119]: Can you also add a caae with braced init.
https://godbolt.org/z/EaY8Ess3v

In that case, this should be valid through lifetime extension.

File: clang/test/SemaCXX/cxx20-warn-dangling-paren-list-agg-init.cpp [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+namespace std {
+template <class T> struct remove_reference { typedef T type; };
+template <class T> struct remove_reference<T&> { typedef T type; };
+template <class T> struct remove_reference<T&&> { typedef T type; };
+
+template <class T> typename remove_reference<T>::type &&move(T &&t);
+} // namespace std
+
+// dcl.init 16.6.2.2
+struct A {
```
Comment (usx95) [inline clang/test/SemaCXX/cxx20-warn-dangling-paren-list-agg-init.cpp:12]: Can you also add more nesting:
https://godbolt.org/z/9jbsxPjrW
https://godbolt.org/z/cj97eaPsK

File: clang/test/SemaCXX/paren-list-agg-init.cpp (hunk 1 above)
Comment (hokein) [reply clang/test/SemaCXX/paren-list-agg-init.cpp:119]: Done. Added one to the `cxx20-warn-dangling-paren-list-agg-init.cpp`.

File: clang/test/SemaCXX/cxx20-warn-dangling-paren-list-agg-init.cpp (hunk 2 above)
Comment (hokein) [reply clang/test/SemaCXX/cxx20-warn-dangling-paren-list-agg-init.cpp:12]: Thanks for the great examples. Added.

**Comment by hokein** [comment]
> LGTM, could you add a summary of the change to the commit message? I know the context is available in the linked issue, but I'd like the gist of it to be preserved in the commit history directly.

Done.

**Comment by usx95** [comment]
Wow. This looks quite nice now without the special handling in `NoExtend` case. 
Was the missing piece `    Path.pop_back();` ?

File: clang/lib/Sema/CheckExprLifetime.cpp [hunk 3]
```diff
@@ -961,6 +963,17 @@ static void visitLocalsRetainedByInitializer(IndirectLocalPath &Path,
...
+  if (auto *CPE = dyn_cast<CXXParenListInitExpr>(Init)) {
+    Path.push_back({IndirectLocalPathEntry::ParenAggInit, CPE});
+    for (auto *I : CPE->getInitExprs()) {
+      if (I->isGLValue())
+        visitLocalsRetainedByReferenceBinding(Path, I, RK_ReferenceBinding,
+                                              Visit);
+      else
+        visitLocalsRetainedByInitializer(Path, I, Visit, true);
+    }
+    Path.pop_back();
```
Comment (usx95) [inline clang/lib/Sema/CheckExprLifetime.cpp:975]: nit: Use `RevertToOldSizeRAII`.

**Review by usx95 — APPROVED** [review]

File: clang/lib/Sema/CheckExprLifetime.cpp (hunk 3 above)
Comment (hokein) [reply clang/lib/Sema/CheckExprLifetime.cpp]: Done.

**Pull request merged**